warehouse/tuf/repository.py

```diff
@@ -0,0 +1,497 @@
+
+        Args:
+            snapshot_expires: New Snapshot expiration datetime
+            snapshot_metadata: Optional, specific Snapshot Metadata
+            store: Optional, Default False, stores in the repository storage
+
+        Returns:
+            Snapshot metadata
+            ``Metadata[Snapshot]``
+        """
+        if snapshot_metadata is None:
+            snapshot_metadata = self.load_role(Snapshot.type)
+
+        snapshot_metadata.signed.version += 1
+        snapshot_metadata.signed.expires = snapshot_expires
+        snapshot_keys = self.key_backend.get(Snapshot.type, "private")
+        for key in snapshot_keys:
+            snapshot_metadata.sign(SSlibSigner(key), append=True)
+
+        if store is True:
+            self._store(Snapshot.type, snapshot_metadata)
+
+        return snapshot_metadata
+
+    def snapshot_update_meta(
+        self,
+        meta_role_name: str,
+        meta_role_version: int,
+        snapshot_metadata: Optional[Metadata[Snapshot]] = None,
+    ) -> Metadata[Snapshot]:
+        """
+        Update the Snapshot Metadata Meta.
+
+        Args:
+            meta_role_name: role name
+            meta_role_version: role version
+            snapshot_metadata: Optional, specific Snapshot Metadata
+
+        Return:
+            Snapshot metadata with updated Meta
+            ``Metadata[Snapshot]``
+        """
+        if snapshot_metadata is None:
+            snapshot_metadata = self.load_role(Snapshot.type)
+
+        snapshot_metadata.signed.meta[f"{meta_role_name}.json"] = MetaFile(
+            version=meta_role_version
+        )
+
+        return snapshot_metadata
+
+    def add_targets(
+        self,
+        payload: Dict[str, List[TargetsPayload]],
+        key_rolename: str,
+    ) -> Metadata[Snapshot]:
+        """
+        Add a list of targets from a payload to the Role Metadata, bumps the
+        role Metadata version, stores, bumps Timestamp and
+        Snapshot Metadata version and store them.
+
+        Args:
+            payload: Dictionary with role names and list of ``TargetPayload``
+            timestamp_expires: New Timestamp expiration datetime
+            snapshot_expires: New Snapshot expiration datetime
+            rolename_key: Key signer rolename
+        """
+        snapshot_metadata = self.load_role(Snapshot.type)
+
+        for rolename, targets in payload.items():
+            role_metadata = self.load_role(rolename)
+            for target in targets:
+                target_file = TargetFile.from_dict(target.fileinfo, target.path)
+                role_metadata.signed.targets[target.path] = target_file
+
+            role_metadata.signed.version += 1
+            role_keys = self.key_backend.get(key_rolename, "private")
+            for key in role_keys:
+                role_metadata.sign(SSlibSigner(key), append=True)
+
+            self._store(rolename, role_metadata)
+            role_metadata = self.bump_role_version(
```

Comment:
> One more thing I noticed wrt interface boundary between `tuf.services` and `tuf.repository`:
> 
> All service methods that update targets metadata, i.e. `init_targets_delegation`, `add_hashed_targets` and `bump_bin_n_roles` transitively bump 'snapshot' and 'timestamp'. But while the latter does this explicitly, the former two do it implicitly by calling methods in `tuf.repository` that do it for them. `bump_snapshot` follows the same principle as `bump_bin_n_roles`, as it transitively but explicitly bumps 'timestamp'.
> 
> I think this makes the 'tuf.repository' interface slightly inconsistent. Is performing the chain of inter-dependent metadata updates a repository task, or is it a task for the application code (service)? I don't say we should change this, but I wanted to take a note about it somewhere.

That was an excellent catch @lukpueh ❤️ 

I removed the implicitly calls for bumping snapshot and timestamp version in the repository (``repository.tuf``).
The only remaining call is for ``bump_role_version`` when adding new targets. For this one, I need to think of a way to avoid degrading the performance.